### DISCUSSION TITLE
Can't override default keymaps.

### DISCUSSION THREAD

**Discussion by hf-xz** [LazyVim/LazyVim#1427] [Q&A] [answered]
I have a simple problem overriding the default window keymap. I want to use `<leader>w` to save file, but `<leader>w` is taken by default window keymap, so I write this in my `keymaps.lua`:

```lua
local function map(mode, lhs, rhs, opts)
  opts = opts or {}
  opts.silent = opts.silent ~= false
  vim.keymap.set(mode, lhs, rhs, opts)
end

-- overwrite windows
vim.keymap.del("n", "<leader>ww")
vim.keymap.del("n", "<leader>wd")
vim.keymap.del("n", "<leader>w-")
vim.keymap.del("n", "<leader>w|")
map("n", "<leader>nw", "<C-W>p", { desc = "Other window" })
map("n", "<leader>nd", "<C-W>c", { desc = "Delete window" })
map("n", "<leader>n-", "<C-W>s", { desc = "Split window below" })
map("n", "<leader>n|", "<C-W>v", { desc = "Split window right" })

-- custom keymaps
map("n", "<leader>h", "<cmd>noh<cr>", { desc = "No highlight" })
map("n", "<leader>w", "<cmd>w<cr>", { desc = "Save file" })
```

And now i get this:

![image](https://github.com/LazyVim/LazyVim/assets/27397614/17d3154e-2183-4563-b374-41a95c6fdf11)

The `<leader>w*` windows actions is gone, but I can't set `<leader>w` for saving and the description of `<leader>n` is not corret. How could I fix this?

**Comment by dpetka2001** [comment, marked as answer]
For the description of your keymaps you should configure `which-key` like so
```lua
{
  "folke/which-key.nvim",
  opts = {
    defaults = {
      ["<leader>n"] = { name = "+windows" },
    },
  },
}
```
The rest should be correct. I tried what you did and `<leader>w` does indeed save the file as you wanted.

**Reply by abeldekat** [reply]
In addition:
The new description of `leader w` only appears when the old `menu` description is unset:

```lua
{
  "folke/which-key.nvim",
  opts = function(_, opts)
      opts.defaults["<leader>w"] = nil
      opts.defaults["<leader>n"] = { name = "+windows" }
      return opts
  end,
}
```

**Comment by hf-xz** [comment]
wow, thanks a lot, this is exactly what i need 🤩

**Comment by Rem0ld** [comment]
Hello, 
I feel something is off with the keymaps. Sometimes the remap is working sometime it doesn't.
I'm struggling with the <leader>w as well. It's like it takes forever for the keymaps to be activated.

I'm wondering if it could be possible to have an options to have LazyVim with default keymaps or no keymaps?
And we choose the keymaps we wanna change/activate/deactivate etc.
At the moment I feel something is off and if I wanna change a keymap I need to remove keymap(s) (<leader>w, means I need to remove 4 or 5 keymaps before I can use it) which makes the process super tedious.
Let me know what you think of this idea.

**Reply by dpetka2001** [reply]
Take a look here in the [docs](https://www.lazyvim.org/configuration#default-settings). You can disable the default keymaps if you'd like and then re-organize everything according to your needs.

**Reply by Rem0ld** [reply]
ahah didn't pay attention to this option.
This distro is clearly amazing, thanks a lot!

**Comment by leikoilja** [comment]
Finally, made it work. 🔥
In my case all i wished for is to map `<leader>ww` to "Save a file". And I am fine to keep all the "defaults" for `<leader>w` group.

After all the tries all i needed to do is to remove `w` mapping first:
```
vim.keymap.del("n", "<leader>w")
vim.keymap.set("n", "<leader>ww", "<cmd>w<CR>", { noremap = true, silent = true, desc = "Save file" })
```
in my `keymaps.lua` and now works as expected. 
No changes needed to `which-key.nvim` config.

**Reply by utkarshgupta137** [reply]
I dug into this & the problem is that LazyVim sets this keymap using `which-key.add` instead of adding a spec to which-key conf. This means that anything described in the spec gets overwritten. I finally figured out the solution:
```lua
vim.keymap.del("n", "<leader>wd")
vim.keymap.del("n", "<leader>wm")
vim.keymap.del("n", "<leader>-")
vim.keymap.del("n", "<leader>|")
vim.keymap.set({ "n", "v", "s" }, "<leader>w", "<cmd>w<cr><esc>", { desc = "Save" })

local wk = require("which-key")
wk.add({
  { "<leader>wm", hidden = true },
})
```
This is different from yours, since I want to use `<leader>w` directly & not use any of `<leader>w` group mappings.

**Reply by leikoilja** [reply]
seem that this got fixed in the latest versions, since whichkey started erroring out for me for `vim.keymap.del("n", "<leader>w")` line. Removing the line fixes the init error, while my `vim.keymap.set("n", "<leader>ww", "<cmd>w<CR>", { noremap = true, silent = true, desc = "Save file" })
` config still works as expected
